src/common/imageio_heif.c

```diff
@@ -0,0 +1,304 @@
+  const size_t width = heif_image_handle_get_width(handle);
+  const size_t height = heif_image_handle_get_height(handle);
+
+  /* Initialize cached image buffer */
+  img->width = width;
+  img->height = height;
+
+  img->buf_dsc.channels = 4;
+  img->buf_dsc.datatype = TYPE_FLOAT;
+  img->buf_dsc.cst = iop_cs_rgb;
+
+  float *mipbuf = (float *)dt_mipmap_cache_alloc(mbuf, img);
+  if (mipbuf == NULL) {
+    dt_print(DT_DEBUG_IMAGEIO,
+             "Failed to allocate mipmap buffer for HEIF image [%s]\n",
+             filename);
+    ret = DT_IMAGEIO_CACHE_FULL;
+    goto out;
+  }
+
+  img->buf_dsc.filters = 0u;
+  img->flags &= ~DT_IMAGE_RAW;
+  img->flags &= ~DT_IMAGE_S_RAW;
+
+  // Get pixel value bit depth (not storage bit depth)
+  int bit_depth = heif_image_get_bits_per_pixel_range(heif_img, heif_channel_interleaved);
+
+  dt_print(DT_DEBUG_IMAGEIO,
+             "Bit depth: '%d' for HEIF image [%s]\n",
+             bit_depth,
+             filename);
+
+  /* This can be LDR or HDR, it depends on the ICC profile. But if bit_depth <= 8 it must be LDR. */
+  if(bit_depth > 8) {
+    img->flags |= DT_IMAGE_HDR;
+  } else {
+    img->flags &= ~DT_IMAGE_HDR;
+  }
+
+  float max_channel_f = (float)((1 << bit_depth) - 1);
+
+  const uint8_t *const restrict in = (const uint8_t *)data;
+
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(mipbuf, width, height, in, rowbytes, max_channel_f) \
+  schedule(simd:static) \
+  collapse(2)
+#endif
+  for (size_t y = 0; y < height; y++)
+  {
+    for (size_t x = 0; x < width; x++)
+    {
+        uint16_t *in_pixel = (uint16_t *)&in[(y * rowbytes) + (3 * sizeof(uint16_t) * x)];
+        float *out_pixel = &mipbuf[(size_t)4 * ((y * width) + x)];
+
+        /* max_channel_f is 1023.0f for 10bit */
+        out_pixel[0] = ((float)in_pixel[0]) * (1.0f / max_channel_f);
+        out_pixel[1] = ((float)in_pixel[1]) * (1.0f / max_channel_f);
+        out_pixel[2] = ((float)in_pixel[2]) * (1.0f / max_channel_f);
+        out_pixel[3] = 0.0f; /* alpha */
+    }
+  }
+  ret = DT_IMAGEIO_OK;
+
+  out:
+  // cleanup handles
+  if(heif_img) {
+    heif_image_release(heif_img);
+  }
+  if(handle) {
+    heif_image_handle_release(handle);
+  }
+  heif_context_free(ctx);
+
+  return ret;
+}
+
+
+int dt_imageio_heif_read_profile(const char *filename,
+                                uint8_t **out,
+                                dt_colorspaces_cicp_t *cicp)
+{
+  /* set default return values */
+  int size = 0;
+  *out = NULL;
+  cicp->color_primaries = heif_color_primaries_unspecified;
+  cicp->transfer_characteristics = heif_transfer_characteristic_unspecified;
+  cicp->matrix_coefficients = heif_matrix_coefficients_unspecified;
+
+  struct heif_image_handle* handle = NULL;
+
+  struct heif_context* ctx = heif_context_alloc();
+  if(!ctx) {
+    dt_print(DT_DEBUG_IMAGEIO,
+             "Unable to allocate HEIF context\n");
+    return DT_IMAGEIO_CACHE_FULL;
```

Comment:
Profile reading no longer returns imageio error codes, just goto out and return 0